docs/03-astropy_dev_env.rst

```diff
@@ -0,0 +1,178 @@
+Use ``git`` for committing your changes and pushing those changes to
+your ``my-code-update`` branch on Github.
+
+Here's an example of creating a new file and pushing it to the new branch:
+
+::
+
+       echo "print('hello world')" > my-new-code.py
+       git add my-new-code.py
+       git commit -m "first commit"
+       git push origin my-code-update
+
+For more help with learning Git and Github, see the `Github
+Help <https://help.github.com/>`__ pages, the `Git and Github Learning
+Resources <https://help.github.com/articles/git-and-github-learning-resources/>`__
+guide, `try.github.io <http://try.github.io/>`_, and the `Github Learning Lab  <https://lab.github.com/>`_
+
+When you are done working on development, don't forget to exit the
+astropy-dev environment.
+
+::
+
+    source deactivate
+
+When you want your changes to be incorporated into Astropy, `submit a pull request on Github <https://help.github.com/articles/creating-a-pull-request/>`__. If you're looking for `quick ways to contribute to Astropy <http://www.astropy.org/contribute.html#code>`__, check out the issues page on the main Astropy Github repo.
+
+
+Keeping your fork up to date
+============================
+
+If it has been awhile since you last forked the Astropy repo, your fork
+will likely need updating. This tutorial will teach you how to use the
+command line to connect to the main Astropy Github repo, update to the
+latest development version, and push those changes to your personal
+Astropy fork on Github.
+
+Step 1: Add the core Astropy repo to your git config file
+---------------------------------------------------------
+
+-  Go to https://github.com/astropy/astropy and click on the green
+   "Clone or Download" button in the upper right-hand corner of the
+   page.
+-  Go to the directory where you have the Astropy repo on your computer,
+   for example:
+
+   ::
+
+       cd ~/astropy-dev
+
+-  Add a new remote named *upstream* to your local copy of the Astropy
+   repo
+
+   ::
+
+       git remote add upstream git@github.com:astropy/astropy.git
+
+Step 2: Pull any changes from the master branch of the main Astropy repo
+------------------------------------------------------------------------
+
+-  Make sure you are in the master branch of your local Astropy repo
+
+   ::
+
+       git checkout master
+
+-  Use ``git pull`` to update your local master branch from the upstream
+   Astropy master branch
+
+   ::
+
+       git pull upstream master
+
+Step 3: Push the changes to your fork on Github
+-----------------------------------------------
+
+-  Use ``git push`` to update your Github fork of Astropy:
+
+   ::
+
+       git push origin master
+   
+   If you've already made some changes to your own master branch, you may need to force the push with the `--force` command. This may cause you to lose some changes or issues with your git history. This is why it's good practice to **always develop in a separate branch**.
```

Comment:
If someone has messed with their local master a lot (which I see from time to time) even ``--force`` doesn't work, and instead the thing to do is:
```
git fetch astropy
git reset --hard astropy/master
```
which is basically equivalent to completely deleting the local master and replacing it with the master on the "astropy" remote (or "upstream" if you decide not to go with my renaming suggestions)